noir-projects/aztec-nr/aztec/src/encrypted_logs/log_assembly_strategies/default_aes128/event.nr

```diff
@@ -0,0 +1,283 @@
+
+    let mut log_bytes = get_arr_of_size__log_bytes__from_PT::<(N * 32) + 32>();
+
+    log_bytes[0] = eph_pk_sign_byte;
+    let mut offset = 1;
+    for i in 0..header_ciphertext_bytes.len() {
+        log_bytes[offset + i] = header_ciphertext_bytes[i];
+    }
+    offset += header_ciphertext_bytes.len();
+
+    for i in 0..ciphertext_bytes.len() {
+        log_bytes[offset + i] = ciphertext_bytes[i];
+    }
+    offset += ciphertext_bytes.len();
+
+    for i in 0..log_bytes_padding_to_mult_31.len() {
+        log_bytes[offset + i] = log_bytes_padding_to_mult_31[i];
+    }
+
+    // *****************************************************************************
+    // Convert bytes back to fields
+    // *****************************************************************************
+
+    let log_bytes_as_fields = be_bytes_31_to_fields(log_bytes);
+
+    // *****************************************************************************
+    // Prepend / append fields, to create the final log
+    // *****************************************************************************
+
+    // In this strategy, we prepend [tag, eph_pk.x]
+
+    // We assume that the sender wants for the recipient to find the tagged note,
+    // and therefore that they will cooperate and use the correct tag. Usage of a bad
+    // tag will result in the recipient not being able to find the note automatically.
+    let tag = unsafe { get_app_tag_as_sender(sender, recipient) };
+    increment_app_tagging_secret_index_as_sender(sender, recipient);
+
+    let mut final_log: [Field; PRIVATE_LOG_SIZE_IN_FIELDS] = [0; PRIVATE_LOG_SIZE_IN_FIELDS];
+
+    final_log[0] = tag;
+    final_log[1] = eph_pk.x;
+
+    let mut offset = 2;
+    for i in 0..log_bytes_as_fields.len() {
+        final_log[offset + i] = log_bytes_as_fields[i];
+    }
+    offset += log_bytes_as_fields.len();
+
+    for i in offset..PRIVATE_LOG_SIZE_IN_FIELDS {
+        final_log[i] = unsafe { random() };
+    }
+
+    final_log
+}
+
+unconstrained fn compute_log_unconstrained<Event, let N: u32>(
+    context: PrivateContext,
+    event: Event,
+    recipient: AztecAddress,
+    sender: AztecAddress,
+) -> [Field; PRIVATE_LOG_SIZE_IN_FIELDS]
+where
+    Event: EventInterface<N>,
+{
+    compute_log(context, event, recipient, sender)
+}
+
+pub fn encode_and_encrypt_event<Event, let N: u32>(
+    context: &mut PrivateContext,
+    recipient: AztecAddress,
+    sender: AztecAddress,
+) -> fn[(&mut PrivateContext, AztecAddress, AztecAddress)](Event) -> ()
+where
+    Event: EventInterface<N>,
+{
+    |e: Event| {
+        let encrypted_log = compute_log(*context, e, recipient, sender);
+        context.emit_private_log(encrypted_log);
+    }
+}
+
+// Important note: this function -- although called "unconstrained" -- the
+// function is not labelled as `unconstrained`, because we pass a reference to the
+// context.
+pub fn encode_and_encrypt_event_unconstrained<Event, let N: u32>(
```

Comment:
It's not because of the ref, it's because the emission of the event _is_ constrained (it must be). What's unconstrained is the content of the event.

Reply:
Well, technically emission of the event (creating a public input) doesn't _need_ to be constrained, if the user is already accepting that the computation of the event doesn't need to be constrained. They could just be free fields in the databus column that aren't constrained to relate to any other fields in the circuit. What does need to be constrained is any incrementation of side effect counters at the time of pushing stuff to the context. I was actually worried for a while that it might be possible to push to the context within unconstrained function by _accident_, because that would be horrendously insecure. But conveniently, the fact that the context is passed by ref, and the fact that refs can't be passed to unconstrained functions, means we're safe.

Reply:
>But conveniently, the fact that the context is passed by ref, and the fact that refs can't be passed to unconstrained functions, means we're safe.

This is not a coincidence, it's not refs that cannot be passed but _mutable_ refs, because you cannot be allowed to mutate objects that originated in constrained land in unconstrained functions. So Noir naturally prevents you from getting this wrong.